src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -39,28 +39,25 @@ internal sealed partial class AssetProvider(Checksum solutionChecksum, SolutionA
         using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksums);
         checksums.Add(checksum);
 
-        using var _2 = PooledDictionary<Checksum, T>.GetInstance(out var results);
-        await this.SynchronizeAssetsAsync(assetPath, checksums, results, cancellationToken).ConfigureAwait(false);
+        var called = false;
+        T? result = default;
+        await this.SynchronizeAssetsAsync<T, VoidResult>(assetPath, checksums, (_, asset, _) =>
+        {
+            Contract.ThrowIfTrue(called);
+            called = true;
```

Comment:
>called = true; [](http://example.com/codeflow?start=12&length=14)

Kinda sucks to allocate the closure for these. What about passing an ArrayBuilder over and just using it to store the single result (and doing the verification based on it's length)

Reply:
indeed.